### PR TITLE
Restore Object API

### PR DESCRIPTION
**Pull request by nguptaopensds**
**What this PR does / why we need it**:
This is the PR for restore API of object

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
```
POST:
http://<host_ip>/v1/s3/<bucket_name>/<object_name>
<RestoreRequest>
   <Days>4</Days> 
   <GlacierJobParameters>
     <Tier>STANDARD</Tier>
   </GlacierJobParameters> 
</RestoreRequest> 

```
**Release note**:

```release-note
```

### CODE REVIEW

**Comment by xing-yang** [comment]
Please add a release note.

File: api/pkg/s3/objectrestore.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+				return
+			} else {
+				s3restore.RestoreState = ActionStateRestoring
+			}
+
+			objectMD.RestoreStatus = &s3restore
+
+			res, err := s.s3Client.UpdateObject(ctx, objectMD)
+			if err != nil {
+				log.Logf("object metadata restore updated successfully and response is : %v\n", res)
```
Comment (rhsakarpos) [inline api/pkg/s3/objectrestore.go]: this is error case...please fix log message

File: api/pkg/s3/datastore/aws/aws.go [hunk 2]
```diff
@@ -165,6 +165,26 @@ func (ad *AwsAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context)
 	return NoError
 }
 
+func (ad *AwsAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+
+	bucket := ad.backend.BucketName
+
+	newObjectKey := object.Bucket + "/" + object.Key
+
+	restoreInput := awss3.RestoreObjectInput{Bucket: &bucket, Key: &newObjectKey}
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/aws/aws.go]: if we see the AWS documentation, this is the full RestoreObjectInput. Why are we not using the RestoreRequest parameter ?
`input := &s3.RestoreObjectInput{
    Bucket: aws.String("examplebucket"),
    Key:    aws.String("archivedobjectkey"),
    RestoreRequest: &s3.RestoreRequest{
        Days: aws.Int64(1),
        GlacierJobParameters: &s3.GlacierJobParameters{
            Tier: aws.String("Expedited"),
        },
    },
}`

File: api/pkg/s3/datastore/aws/aws.go [hunk 3]
```diff
@@ -165,6 +165,26 @@ func (ad *AwsAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context)
...
+
+	svc := awss3.New(ad.session)
+	_, err := svc.RestoreObject(&restoreInput)
+	if err != nil {
+		log.Logf("restore object failed, err:%v\n", err)
+		return InternalError
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/aws/aws.go]: why are we returning InternalError ? please return AWS specifc error, like this
`if err != nil {
    if aerr, ok := err.(awserr.Error); ok {
        switch aerr.Code() {
        case s3.ErrCodeObjectAlreadyInActiveTierError:
            fmt.Println(s3.ErrCodeObjectAlreadyInActiveTierError, aerr.Error())
        default:
            fmt.Println(aerr.Error())
        }
    } else {
        // Print the error, cast err to awserr.Error to get the Code and
        // Message from an error.
        fmt.Println(err.Error())
    }
    return
}`

File: api/pkg/s3/datastore/aws/aws.go [hunk 4]
```diff
@@ -165,6 +165,26 @@ func (ad *AwsAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context)
...
+func (ad *AwsAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+
+	bucket := ad.backend.BucketName
+
+	newObjectKey := object.Bucket + "/" + object.Key
+
+	restoreInput := awss3.RestoreObjectInput{Bucket: &bucket, Key: &newObjectKey}
+
+	svc := awss3.New(ad.session)
+	_, err := svc.RestoreObject(&restoreInput)
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/aws/aws.go]: why are we ignoring the important information in the first response parameter ?
`type RestoreObjectOutput struct {

    // If present, indicates that the requester was successfully charged for the
    // request.
    RequestCharged *string `location:"header" locationName:"x-amz-request-charged" type:"string" enum:"RequestCharged"`

    // Indicates the path in the provided S3 output location where Select results
    // will be restored to.
    RestoreOutputPath *string `location:"header" locationName:"x-amz-restore-output-path" type:"string"`
    // contains filtered or unexported fields
}`

File: api/pkg/s3/datastore/azure/azure.go [hunk 5]
```diff
@@ -177,6 +177,23 @@ func (ad *AzureAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context
...
 }
+
+func (ad *AzureAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+	newObjectKey := object.Bucket + "/" + object.Key
+	blobURL := ad.containerURL.NewBlockBlobURL(newObjectKey)
+	setTierRsp, err := blobURL.SetTier(ctx, azblob.AccessTierHot, azblob.LeaseAccessConditions{})
+	log.Logf("blobURL=%v,err=%v\n", blobURL, err)
+	if err != nil {
+		log.Logf("[AzureAdapter] post restore object failed:%v\n", err)
+		return S3Error{Code: 500, Description: "post failed"}
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/azure/azure.go]: please handle specific errors from azure

File: api/pkg/s3/datastore/azure/azure.go [hunk 6]
```diff
@@ -177,6 +177,23 @@ func (ad *AzureAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context
 	}
 	return NoError
 }
+
+func (ad *AzureAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+	newObjectKey := object.Bucket + "/" + object.Key
+	blobURL := ad.containerURL.NewBlockBlobURL(newObjectKey)
+	setTierRsp, err := blobURL.SetTier(ctx, azblob.AccessTierHot, azblob.LeaseAccessConditions{})
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/azure/azure.go]: Azure API doc says this...how are we ensuring the lease conditions match ?

> By setting lease access conditions, requests will fail if the provided lease does not match the active lease on the blob.

File: api/pkg/s3/datastore/azure/azure.go [hunk 7]
```diff
@@ -177,6 +177,23 @@ func (ad *AzureAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context
...
+	log.Logf("blobURL=%v,err=%v\n", blobURL, err)
+	if err != nil {
+		log.Logf("[AzureAdapter] post restore object failed:%v\n", err)
+		return S3Error{Code: 500, Description: "post failed"}
+	}
+	if setTierRsp.StatusCode() != http.StatusOK && setTierRsp.StatusCode() != http.StatusAccepted {
+		log.Logf("[AzureAdapter] restore failed, status code:%d\n", setTierRsp.StatusCode())
+		return S3Error{Code: 500, Description: "restore failed"}
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/azure/azure.go]: specific errors please

File: api/pkg/s3/datastore/azure/azure.go [hunk 8]
```diff
@@ -177,6 +177,23 @@ func (ad *AzureAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context
 	}
 	return NoError
 }
+
+func (ad *AzureAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+	newObjectKey := object.Bucket + "/" + object.Key
+	blobURL := ad.containerURL.NewBlockBlobURL(newObjectKey)
+	setTierRsp, err := blobURL.SetTier(ctx, azblob.AccessTierHot, azblob.LeaseAccessConditions{})
+	log.Logf("blobURL=%v,err=%v\n", blobURL, err)
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/azure/azure.go]: in any case of successful restore, where is our (SODA) database being updated ?

File: api/pkg/s3/datastore/ceph/ceph.go [hunk 9]
```diff
@@ -141,6 +141,10 @@ func (ad *CephAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context)
 	return NoError
 }
 
+func (ad *CephAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+	return NoError
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/ceph/ceph.go]: please return Not Supported wherever this operation is not supported (CEPH, GCP)

File: api/pkg/s3/datastore/hws/hws.go [hunk 10]
```diff
@@ -113,6 +113,21 @@ func (ad *OBSAdapter) DELETE(object *pb.DeleteObjectInput, ctx context.Context)
 	return NoError
 }
 
+func (ad *OBSAdapter) RestoreObject(object *pb.RestoreObjectInput, ctx context.Context) S3Error {
+
+	newObjectKey := object.Bucket + "/" + object.Key
+	restoreObjectInput := obs.RestoreObjectInput{Bucket: ad.backend.BucketName, Key: newObjectKey, Days: int(object.Days), Tier: obs.RestoreTierType(object.Tier)}
+	log.Logf("In huawei obs database and input is :", restoreObjectInput)
```
Comment (rhsakarpos) [inline api/pkg/s3/datastore/hws/hws.go]: company name mentioned here, is okay ?

**Comment by nguptaopensds** [comment]
Closing the PR because the context have been changed

**Pull request closed without merging**